### PR TITLE
Enhance the stability of Asterinas Benchmark

### PR DESCRIPTION
**Pull request by grief8**
This PR enhances the stability of benchmark CI:

1. `bench_linux_and_aster.sh` : `make run BENCHMARK=${benchmark} ENABLE_KVM=1 RELEASE=1 2>&1 | tee "${aster_output}"` cannot return a non-zero exit code, which affects the retry attempt of the action `nick-invision/retry@v2`.
2. `benchmark_asterinas.yml`: 
      - Schedule to run on every day at 20:00 UTC (04:00 UTC +8)
      - Add `make --no-print-directory -C test/apps/mongoose clean` before retrying commands to ensure that `mongoose` headers are deleted in case of empty headers.

Note: Another way to fix it is adding a header check to `test/apps/mongoose`, which forces downloading if the headers are empty. However, it doesn't seem elegant. 
```makefile
# SPDX-License-Identifier: MPL-2.0

CUR_DIR := $(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
BUILD_DIR := $(CUR_DIR)/../../build/initramfs/test/network
MONGOOSE_DIR := $(CUR_DIR)
MONGOOSE_C := $(MONGOOSE_DIR)/mongoose.c
MONGOOSE_H := $(MONGOOSE_DIR)/mongoose.h
MONGOOSE_FILES := $(MONGOOSE_C) $(MONGOOSE_H)
MONGOOSE_O := $(MONGOOSE_DIR)/mongoose.o
SERVER_C := http_server.c
SERVER_BIN := $(BUILD_DIR)/http_server
CLIENT_C := http_client.c
CLIENT_BIN := $(BUILD_DIR)/http_client
BINS := $(SERVER_BIN) $(CLIENT_BIN)
CC := cc
CFLAGS := -W -Wall -Wextra -g -I. -I$(MONGOOSE_DIR) -DMG_ENABLE_LINES=1

.PHONY: all 
all: check-mongoose-files $(BINS) 

$(SERVER_BIN): $(SERVER_C) $(MONGOOSE_O) | $(BUILD_DIR)
	$(CC) $^ $(CFLAGS) -o $@

$(CLIENT_BIN): $(CLIENT_C) $(MONGOOSE_O) | $(BUILD_DIR)
	$(CC) $^ $(CFLAGS) -o $@

$(MONGOOSE_O): $(MONGOOSE_FILES) 
	$(CC) -c $(MONGOOSE_C) $(CFLAGS) -o $@

check-mongoose-files: | $(MONGOOSE_DIR)
	@for file in $(MONGOOSE_FILES); do \
		if [ ! -s $$file ]; then \
			url="https://raw.githubusercontent.com/cesanta/mongoose/7.13/$$(basename $$file)"; \
			echo "Downloading $$url to $$file"; \
			wget -O $$file "$$url" || { echo "Failed to download $$url"; exit 1; }; \
		fi; \
	done

$(BUILD_DIR) $(MONGOOSE_DIR):
	@mkdir -p $@

.PHONY: clean
clean:
	@rm -f $(BINS) $(MONGOOSE_O) $(MONGOOSE_FILES)
```

### CODE REVIEW

File: .github/workflows/benchmark_asterinas.yml [hunk 1]
```diff
@@ -41,7 +41,9 @@ jobs:
         command: |
           make install_osdk
           bash test/benchmark/bench_linux_and_aster.sh ${{ matrix.benchmark }}
-        on_retry_command: make clean
+        on_retry_command: |
+          make clean
+          make --no-print-directory -C test/apps/mongoose clean
```
Comment (tatetian) [inline .github/workflows/benchmark_asterinas.yml:46]: First of all, I don't think `make clean` is desirable. Compiling the kernel is really time consuming. Having to `make clean` on retry means we are going to repeat a lot of efforts. And it means our Makefile is not written in a robust way. Ideally, if anything bad happens during `make install_osdk` or `bench_linux_and_aster.sh`. We should be able to simply rerun `make install_osdk` or `bench_linux_and_aster.sh`.

Second, I don't know why `mongoose` test has to be so special that the benchmark CI code has to single it out. The benchmark-related code should not be concerned with anything particular tests. If the `mongoose` test cannot be re-run on retry, then it means the `mongoose` test is poorly written and should be fixed.

Comment (grief8) [reply .github/workflows/benchmark_asterinas.yml:46]: Our makefile is indeed not perfect and lacks a logic to verify whether the download is successful. For example, when `vdso64.so` was downloaded in real time before, `vdso64.so` would be downloaded through wget every time it was compiled. If this compilation fails by accident, the makefile will not re-download it next time because the `vdso64.so` file exists, causing subsequent compilation failures.

I hope that no residual files will be left before the download process is completed, so as not to affect the judgment of the makefile. Therefore, adding a temporary file is more reasonable.
```makefile
$(MONGOOSE_FILES): | $(MONGOOSE_DIR)
@set -e;
wget -c -O $@.tmp "https://raw.githubusercontent.com/cesanta/mongoose/7.13/$(notdir $@)"
@mv $@.tmp $@
```
The above code will not generate `$(MONGOOSE_FILES)` if the download process times out or is unexpectedly interrupted.

File: test/benchmark/bench_linux_and_aster.sh [hunk 2]
```diff
@@ -61,8 +62,8 @@ run_benchmark() {
     if [ ! -f "${LINUX_KERNEL}" ]; then
         echo "Downloading the Linux kernel image..."
         mkdir -p "${KERNEL_DIR}"
-        curl -L -o "${LINUX_KERNEL}" \
-            -H "Accept: application/vnd.github.v3.raw" \
+        wget --delete-after --header="Accept: application/vnd.github.v3.raw" \
```
Comment (tatetian) [inline test/benchmark/bench_linux_and_aster.sh:65]: From wget man page:

> --delete-after
> This option tells Wget to delete every single file it downloads, after having done so.

It seems that you have misunderstood this `--delete-after` option.

Comment (grief8) [reply test/benchmark/bench_linux_and_aster.sh:65]: I think this explanation may lead to misunderstanding. If you run `wget --delete-after -O mongoose.c "https://raw.githubusercontent.com/cesanta/mongoose/7.13/mongoose.c"`, you will see that `mongoose.c.tmp` is downloaded and moved to `mongoose.c` once succeed. 
image

Comment (tatetian) [reply test/benchmark/bench_linux_and_aster.sh:65]: No, the command `wget --delete-after -O <output_file> <url>` does not work as you intended. 

```
--delete-after
This option tells Wget to delete every single file it downloads, after having done so.
It is useful for pre-fetching popular pages through a proxy, e.g.:
```

The description above has clearly stated that the downloaded file will be deleted always even when the download is successful. Why do we want to _delete_ the file even when the file is downloaded _successfully_?

You are right that `wget --delete-after -O <output_file> <url>` _can_ download the remote resource as `<output_file>`. But that is NOT the point. We can simply use `wget <url>` to download the file, without any arguments. The advanced form of `wget --delete-after -O <output_file> <url>` is supposed to satisfy the following conditions:

1. If the whole file is downloaded, then the file can be found locally;
2. If the file is not downloaded at all or downloaded partially, then the file should not be created.

But `wget --delete-after -O <output_file> <url>` does not work this way.

```
-O file
Use of -O is not intended to mean simply "use the name file instead of the one in the URL ;" rather, it is analogous to shell redirection: wget -O file http://foo is intended to work like wget -O - http://foo > file; file will be truncated immediately, and all downloaded content will be written there.
```

The description above is pretty clear: it works like a redirection. So before the download begins, wget will create a new file as the destination of shell redirection. As the content is downloaded, wget will write the content into both `mongoose.c.tmp` and `mongoose.c`. When the download is complete, the `mongoose.c.tmp` will be deleted, leaving only `mongoose.c`.

The explanation above is supported by experiments. I used CTRL+C to stop wget while it was downloading the remote file. Guess what? An empty `mongoose.c` file remained---as expected. 

image

Comment (grief8) [reply test/benchmark/bench_linux_and_aster.sh:65]: Thanks for the detailed explanation, @tatetian. It's clear that the --delete-after option combined with -O <output_file> doesn't achieve the desired behavior. The intended use case seems to be more about temporary downloads rather than preserving the file post-download.

For our specific needs, it looks like we'll need to explore alternative approaches or scripts to ensure that the file is only retained if the download is complete even CTRL+C is pressed. Therefore, I think using `TEMP_FILE` is the simplest way.  We download them to `/tmp` and use `wget --continue` to resume transmission after a break.
```makefile
$(MONGOOSE_FILES): | $(MONGOOSE_DIR)
	@if [ ! -d "/tmp" ]; then \
		sudo mkdir /tmp && \
		sudo chmod 1777 /tmp || { \
			echo "Error: Failed to create /tmp directory."; \
			exit 1; \
		}; \
	fi && \
	TARGET_FILE=$(notdir $@) && \
	TEMP_FILE="/tmp/$$TARGET_FILE" && \
	wget --continue -O "$$TEMP_FILE" "https://raw.githubusercontent.com/cesanta/mongoose/7.13/$$TARGET_FILE" && \
	mv "$$TEMP_FILE" "$@" 
```

Thanks again for pointing out the nuances of the command. It's always helpful to get a deeper understanding of how these tools work under the hood.

File: test/Makefile [hunk 3]
```diff
@@ -51,7 +51,17 @@ $(INITRAMFS)/lib/x86_64-linux-gnu: | $(VDSO_LIB)
 
 $(VDSO_LIB): | $(VDSO_DIR)
 	@# TODO: use a custom compiled vdso.so file in the future.
-	@wget https://raw.githubusercontent.com/asterinas/linux_vdso/2a6d2db/vdso64.so -O $@ 
+	@if [ ! -d "/tmp" ]; then \
+		sudo mkdir /tmp && \
+		sudo chmod 1777 /tmp || { \
```
Comment (tatetian) [inline test/Makefile:56]: I don't understand why we can't assume that `/tmp` exists. Isn't it a standard directory on any Linux machine?

File: test/apps/mongoose/Makefile [hunk 4]
```diff
@@ -28,7 +28,17 @@ $(MONGOOSE_O): $(MONGOOSE_FILES)
...
+		sudo mkdir /tmp && \
+		sudo chmod 1777 /tmp || { \
+			echo "Error: Failed to create /tmp directory."; \
+			exit 1; \
+		}; \
+	fi && \
+	TARGET_FILE=$(notdir $@) && \
+	TEMP_FILE="/tmp/$$TARGET_FILE" && \
+	wget --continue -O "$$TEMP_FILE" "https://raw.githubusercontent.com/cesanta/mongoose/7.13/$$TARGET_FILE" && \
+	mv "$$TEMP_FILE" "$@"
```
Comment (tatetian) [inline test/apps/mongoose/Makefile:41]: Do not repeat yourself. Wrap the logic above as a reusable script named `atomic_wget`. Add it to the `tools/` directory. And use the script in all build and CI code.

File: test/Makefile (hunk 3 above)
Comment (grief8) [reply test/Makefile:56]: I believe unix-like systems have `/tmp` unless the administrator customizes the system, so it makes sense to remove the check.

File: test/apps/mongoose/Makefile (hunk 4 above)
Comment (grief8) [reply test/apps/mongoose/Makefile:41]: This is much better. Thank you for your instruction!

File: tools/atomic_wget.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+wget --continue $WGET_OPTIONS -O "$TEMP_FILE" "$2" || { \
+    echo "Error: Failed to download $1."; \
+    exit 1; \
+}
+
+mv "$TEMP_FILE" "$1" || { \
+    echo "Error: Failed to move $TEMP_FILE to $1."; \
+    exit 1; \
+}
```
Comment (tatetian) [inline tools/atomic_wget.sh:23]: This script is not written in a production-grade quality. I ask ChatGPT to revise it and then modify a few locations manually. See the revised version.

```bash
#!/bin/bash

# SPDX-License-Identifier: MPL-2.0

# Ensure the script is invoked with the necessary arguments
if [ "$#" -lt 2 ]; then
    echo "Atomic File Download with wget."
    echo "This script downloads a file from a specified URL using wget. \
        The downloaded file will appear locally as a complete file or not appear at all in case of an error."
    echo ""
    echo "Usage: $0 <output_file> <url> [wget_options]"
    echo ""
    echo "Arguments:"
    echo "  <output_file>   Name of the file to save the downloaded content."
    echo "  <url>           URL of the file to be downloaded."
    echo "  [wget_options]  Optional: Additional wget options for customized download behavior."
    exit 1
fi

OUTPUT_FILE="$1"
URL="$2"
WGET_OPTIONS="${3:-}"

TEMP_FILE=/tmp/$(mktemp)

# Download the file using wget
if ! wget $WGET_OPTIONS -O "$TEMP_FILE" "$URL"; then
    echo "Error: Failed to download $URL."
    rm -f "$TEMP_FILE"
    exit 1
fi

# Move the temporary file to the target location
if ! mv "$TEMP_FILE" "$OUTPUT_FILE"; then
    echo "Error: Failed to move $TEMP_FILE to $OUTPUT_FILE."
    rm -f "$TEMP_FILE"
    exit 1
fi

echo "File downloaded successfully to $OUTPUT_FILE."
```

Improvements made:
* Input Validation: Added a check to ensure the script is invoked with the necessary arguments.
* Temporary File Handling: Used mktemp to create a secure temporary file.
* Conditional Check and Cleanup: Enhanced error handling to remove the temporary file if downloading or moving the file fails.
* Readability: Improved variable names and added explanatory comments for better readability.
* Optional Arguments Handling: Simplified handling of optional wget options using parameter expansion.

File: tools/atomic_wget.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+TEMP_FILE="/tmp/$TARGET_FILE"
+
+# Check if $3 is provided and not empty
+if [ -n "$3" ]; then
+    WGET_OPTIONS="$3"
+else
+    WGET_OPTIONS=""
+fi
+
+wget --continue $WGET_OPTIONS -O "$TEMP_FILE" "$2" || { \
```
Comment (tatetian) [inline tools/atomic_wget.sh:15]: I don't think `--continue` will have any effect when `-O` is also given.

File: tools/atomic_wget.sh (hunk 5 above)
Comment (grief8) [reply tools/atomic_wget.sh:23]: Thank you for your patient instructions! I will check the script in detail before the next commit.

File: tools/atomic_wget.sh (hunk 6 above)
Comment (grief8) [reply tools/atomic_wget.sh:15]: Not really. I have tried the intermittent transfers before and found that it indeed takes effect. Here is an example.
image
